### ISSUE TITLE
Optional federated type

### ISSUE THREAD

**Issue by kamilglod** [mirumee/ariadne#491]
When I want to return a type that has required fields **or** None unexpected error is thrown 

Here https://github.com/mirumee/ariadne/blob/0.12.0/tests/federation/test_schema.py#L431-L469 you have a test that works correctly. However, when you'll change type fields to the required test will fail with error:
```
[GraphQLError('Cannot return null for non-nullable field Product.name.', locations=[SourceLocation(line=6, column=25)], path=['_entities', 0, 'name'])]
```

Test to reproduce:
```python
def test_federated_schema_execute_reference_resolver_that_returns_none():
    type_defs = """
        type Query {
            rootField: String
        }
        type Product @key(fields: "upc") {
            upc: Int!
            name: String!
        }
    """

    product = FederatedObjectType("Product")

    @product.reference_resolver()
    def product_reference_resolver(_obj, _info, reference):
        assert reference["upc"] == 1
        # return None

    schema = make_federated_schema(type_defs, product)

    result = graphql_sync(
        schema,
        """
            query GetEntities($representations: [_Any!]!) {
                _entities(representations: $representations) {
                    ... on Product {
                        __typename
                        name
                    }
                }
            }
        """,
        variable_values={"representations": [{"__typename": "Product", "upc": 1,},],},
    )

    assert result.errors is None
    assert result.data["_entities"][0]["__typename"] == "Product"
    assert result.data["_entities"][0]["name"] is None
```

**Comment by rafalp** [comment]
I'm not sure I see a problem here?

If field is marked with `!` its declared as non-nullable, and query executor will raise an error when it's resolved to null value, as per GraphQL spec. This test will fail after changing type definition because data in question is not given for product to resolve from in this test.

**Comment by kamilglod** [comment]
Understood, but it works like that for normal resolvers - when you return optional type like `product: Product` than you can return `None` or `Product` object. Here there is no option to declare that you want to return `Product` or `None` but you can return `None` if all fields of `Product` are optional. 
GraphQL spec says that you cannot return required field of type as `None`, but you can return None instead of type if whole type is optional.

**Comment by rafalp** [comment]
Ah, I see! Are you testing this on Ariadne 0.12? I remember there was a bug that sounds much like what you are describing with fix being committed and released today as part of Ariadne 0.13

**Comment by kamilglod** [comment]
Indeed, master branch version works correctly! :) 
When you're planning to release 0.13?

**Comment by rafalp** [comment]
@kamilglod it was released 4.5 hours ago, enjoy!
